File: java/Apparat.Core/src/com/joa_ebert/apparat/abc/bytecode/BytecodeDecoder.java
/*
 * This file is part of Apparat.
 * 
 * Apparat is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Apparat is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with Apparat. If not, see <http://www.gnu.org/licenses/>.
 * 
 * Copyright (C) 2009 Joa Ebert
 * http://www.joa-ebert.com/
 * 
 */

package com.joa_ebert.apparat.abc.bytecode;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;

import com.joa_ebert.apparat.abc.AbcEnvironment;
import com.joa_ebert.apparat.abc.ConstantPool;
import com.joa_ebert.apparat.abc.bytecode.analysis.BytecodePrinter;
import com.joa_ebert.apparat.abc.bytecode.operations.ApplyType;
import com.joa_ebert.apparat.abc.bytecode.operations.AsType;
import com.joa_ebert.apparat.abc.bytecode.operations.Call;
import com.joa_ebert.apparat.abc.bytecode.operations.CallMethod;
import com.joa_ebert.apparat.abc.bytecode.operations.CallPropLex;
import com.joa_ebert.apparat.abc.bytecode.operations.CallPropVoid;
import com.joa_ebert.apparat.abc.bytecode.operations.CallProperty;
import com.joa_ebert.apparat.abc.bytecode.operations.CallStatic;
import com.joa_ebert.apparat.abc.bytecode.operations.CallSuper;
import com.joa_ebert.apparat.abc.bytecode.operations.CallSuperVoid;
import com.joa_ebert.apparat.abc.bytecode.operations.Coerce;
import com.joa_ebert.apparat.abc.bytecode.operations.Construct;
import com.joa_ebert.apparat.abc.bytecode.operations.ConstructProp;
import com.joa_ebert.apparat.abc.bytecode.operations.ConstructSuper;
import com.joa_ebert.apparat.abc.bytecode.operations.Debug;
import com.joa_ebert.apparat.abc.bytecode.operations.DebugFile;
import com.joa_ebert.apparat.abc.bytecode.operations.DebugLine;
import com.joa_ebert.apparat.abc.bytecode.operations.DecLocal;
import com.joa_ebert.apparat.abc.bytecode.operations.DecLocalInt;
import com.joa_ebert.apparat.abc.bytecode.operations.DefaultXmlNamespace;
import com.joa_ebert.apparat.abc.bytecode.operations.DeleteProperty;
import com.joa_ebert.apparat.abc.bytecode.operations.FindPropStrict;
import com.joa_ebert.apparat.abc.bytecode.operations.FindProperty;
import com.joa_ebert.apparat.abc.bytecode.operations.GetDescendants;
import com.joa_ebert.apparat.abc.bytecode.operations.GetGlobalSlot;
import com.joa_ebert.apparat.abc.bytecode.operations.GetLex;
import com.joa_ebert.apparat.abc.bytecode.operations.GetLocal;
import com.joa_ebert.apparat.abc.bytecode.operations.GetProperty;
import com.joa_ebert.apparat.abc.bytecode.operations.GetScopeObject;
import com.joa_ebert.apparat.abc.bytecode.operations.GetSlot;
import com.joa_ebert.apparat.abc.bytecode.operations.GetSuper;
import com.joa_ebert.apparat.abc.bytecode.operations.HasNext2;
import com.joa_ebert.apparat.abc.bytecode.operations.IncLocal;
import com.joa_ebert.apparat.abc.bytecode.operations.IncLocalInt;
import com.joa_ebert.apparat.abc.bytecode.operations.InitProperty;
import com.joa_ebert.apparat.abc.bytecode.operations.IsType;
import com.joa_ebert.apparat.abc.bytecode.operations.Jump;
import com.joa_ebert.apparat.abc.bytecode.operations.Kill;
import com.joa_ebert.apparat.abc.bytecode.operations.LookupSwitch;
import com.joa_ebert.apparat.abc.bytecode.operations.NewArray;
import com.joa_ebert.apparat.abc.bytecode.operations.NewCatch;
import com.joa_ebert.apparat.abc.bytecode.operations.NewClass;
import com.joa_ebert.apparat.abc.bytecode.operations.NewFunction;
import com.joa_ebert.apparat.abc.bytecode.operations.NewObject;
import com.joa_ebert.apparat.abc.bytecode.operations.PushByte;
import com.joa_ebert.apparat.abc.bytecode.operations.PushDouble;
import com.joa_ebert.apparat.abc.bytecode.operations.PushInt;
import com.joa_ebert.apparat.abc.bytecode.operations.PushNamespace;
import com.joa_ebert.apparat.abc.bytecode.operations.PushShort;
import com.joa_ebert.apparat.abc.bytecode.operations.PushString;
import com.joa_ebert.apparat.abc.bytecode.operations.PushUInt;
import com.joa_ebert.apparat.abc.bytecode.operations.SetGlobalSlot;
import com.joa_ebert.apparat.abc.bytecode.operations.SetLocal;
import com.joa_ebert.apparat.abc.bytecode.operations.SetProperty;
import com.joa_ebert.apparat.abc.bytecode.operations.SetSlot;
import com.joa_ebert.apparat.abc.bytecode.operations.SetSuper;
import com.joa_ebert.apparat.abc.io.AbcInputStream;

/**
 * 
 * @author Joa Ebert
 * 
 */
public final class BytecodeDecoder
{
	private static final OperationFactory factory = new OperationFactory();

	private final ConstantPool pool;

	public BytecodeDecoder( final ConstantPool constantPool )
	{
		this.pool = constantPool;
	}

	public Bytecode decode( final AbcInputStream input ) throws IOException,
			MarkerException
	{
		final Bytecode result = new Bytecode();
		final MarkerManager markers = result.markers;

		while( input.available() > 0 )
		{
			final int position = (int)input.getPosition();
			final int code = input.read();
			final AbstractOperation operation = factory.create( code );

			if( null == operation )
			{
				continue;
			}

			operation.position = position;

			switch( code )
			{
				case Op.AsType:
					( (AsType)operation ).type = pool.getMultiname( input
							.readU30() );
					break;

				case Op.ApplyType:
					( (ApplyType)operation ).typeSize = input.readU30();
					break;

				case Op.Call:
					( (Call)operation ).numArguments = input.readU30();
					break;

				case Op.CallMethod:
					( (CallMethod)operation ).methodIndex = input.readU30();
					( (CallMethod)operation ).numArguments = input.readU30();
					break;

				case Op.CallProperty:
					( (CallProperty)operation ).property = pool
							.getMultiname( input.readU30() );
					( (CallProperty)operation ).numArguments = input.readU30();
					break;

				case Op.CallPropLex:
					( (CallPropLex)operation ).property = pool
							.getMultiname( input.readU30() );
					( (CallPropLex)operation ).numArguments = input.readU30();
					break;

				case Op.CallPropVoid:
					( (CallPropVoid)operation ).property = pool
							.getMultiname( input.readU30() );
					( (CallPropVoid)operation ).numArguments = input.readU30();
					break;

				case Op.CallStatic:
					( (CallStatic)operation ).methodIndex = input.readU30();
					( (CallStatic)operation ).numArguments = input.readU30();
					break;

				case Op.CallSuper:
					( (CallSuper)operation ).name = pool.getMultiname( input
							.readU30() );
					( (CallSuper)operation ).numArguments = input.readU30();
					break;

				case Op.CallSuperVoid:
					( (CallSuperVoid)operation ).name = pool
							.getMultiname( input.readU30() );
					( (CallSuperVoid)operation ).numArguments = input.readU30();
					break;

				case Op.Coerce:
					( (Coerce)operation ).type = pool.getMultiname( input
							.readU30() );
					break;

				case Op.Construct:
					( (Construct)operation ).numArguments = input.readU30();
					break;

				case Op.ConstructProp:
					( (ConstructProp)operation ).property = pool
							.getMultiname( input.readU30() );
					( (ConstructProp)operation ).numArguments = input.readU30();
					break;

				case Op.ConstructSuper:
					( (ConstructSuper)operation ).numArguments = input
							.readU30();
					break;

				case Op.Debug:
					final Debug debugOperation = (Debug)operation;
					debugOperation.type = input.readU08();
					debugOperation.name = pool.getString( input.readU30() );
					debugOperation.register = input.readU08();
					input.readU30();// unused
					break;

				case Op.DebugFile:
					( (DebugFile)operation ).fileName = pool.getString( input
							.readU30() );
					break;

				case Op.DebugLine:
					( (DebugLine)operation ).lineNumber = input.readU30();
					break;

				case Op.DecLocal:
					( (DecLocal)operation ).register = input.readU30();
					break;

				case Op.DecLocalInt:
					( (DecLocalInt)operation ).register = input.readU30();
					break;

				case Op.DeleteProperty:
					( (DeleteProperty)operation ).property = pool
							.getMultiname( input.readU30() );
					break;

				case Op.DefaultXmlNamespace:
					( (DefaultXmlNamespace)operation ).uri = pool
							.getString( input.readU30() );
					break;

				case Op.FindProperty:
					( (FindProperty)operation ).property = pool
							.getMultiname( input.readU30() );
					break;

				case Op.FindPropStrict:
					( (FindPropStrict)operation ).property = pool
							.getMultiname( input.readU30() );
					break;

				case Op.GetDescendants:
					( (GetDescendants)operation ).name = pool
							.getMultiname( input.readU30() );
					break;

				case Op.GetGlobalSlot:
					( (GetGlobalSlot)operation ).slotIndex = input.readU30();
					break;

				case Op.GetLex:
					( (GetLex)operation ).property = pool.getMultiname( input
							.readU30() );
					break;

				case Op.GetLocal:
					( (GetLocal)operation ).register = input.readU30();
					break;

				case Op.GetProperty:
					( (GetProperty)operation ).property = pool
							.getMultiname( input.readU30() );
					break;

				case Op.GetScopeObject:
					( (GetScopeObject)operation ).scopeIndex = input.readU08();
					break;

				case Op.GetSlot:
					( (GetSlot)operation ).slotIndex = input.readU30();
					break;

				case Op.GetSuper:
					( (GetSuper)operation ).property = pool.getMultiname( input
							.readU30() );
					break;

				case Op.HasNext2:
					( (HasNext2)operation ).objectRegister = input.readU32();
					( (HasNext2)operation ).indexRegister = input.readU32();
					break;

				case Op.IfEqual:
				case Op.IfFalse:
				case Op.IfGreaterEqual:
				case Op.IfGreaterThan:
				case Op.IfLessEqual:
				case Op.IfLessThan:
				case Op.IfNotEqual:
				case Op.IfNotGreaterEqual:
				case Op.IfNotGreaterThan:
				case Op.IfNotLessEqual:
				case Op.IfNotLessThan:
				case Op.IfStrictEqual:
				case Op.IfStrictNotEqual:
				case Op.IfTrue:
				case Op.Jump:
					//
					// The position advanced 4 bytes.
					//
					// The opcode is 1 byte, the S24 is 3 bytes.
					// Therefore we have to add 4 to the current position in
					// order to get the correct marker AFTER those bytes have
					// been consumed.
					//
					( (Jump)operation ).marker = markers.putMarkerAt( position
							+ 0x04 + input.readS24() );
					break;

				case Op.IncLocal:
					( (IncLocal)operation ).register = input.readU30();
					break;

				case Op.IncLocalInt:
					( (IncLocalInt)operation ).register = input.readU30();
					break;

				case Op.InitProperty:
					( (InitProperty)operation ).property = pool
							.getMultiname( input.readU30() );
					break;

				case Op.IsType:
					( (IsType)operation ).type = pool.getMultiname( input
							.readU30() );
					break;

				case Op.Kill:
					( (Kill)operation ).register = input.readU30();
					break;

				case Op.Label:
					markers.putMarkerAt( position );
					break;

				case Op.LookupSwitch:
					final LookupSwitch lookupSwitch = (LookupSwitch)operation;

					lookupSwitch.defaultMarker = markers.putMarkerAt( position
							+ input.readS24() );

					final int caseCount = input.readU30() + 1;

					lookupSwitch.caseMarkers = new ArrayList<Marker>( caseCount );

					for( int i = 0; i < caseCount; ++i )
					{
						lookupSwitch.caseMarkers.add( markers
								.putMarkerAt( position + input.readS24() ) );
					}
					break;

				case Op.NewArray:
					( (NewArray)operation ).numArguments = input.readU30();
					break;

				case Op.NewCatch:
					( (NewCatch)operation ).exceptionHandlerIndex = input
							.readU30();
					break;

				case Op.NewClass:
					( (NewClass)operation ).classIndex = input.readU30();
					break;

				case Op.NewFunction:
					( (NewFunction)operation ).functionIndex = input.readU30();
					break;

				case Op.NewObject:
					( (NewObject)operation ).numProperties = input.readU30();
					break;

				case Op.PushByte:
					( (PushByte)operation ).value = input.readU08();
					break;

				case Op.PushDouble:
					( (PushDouble)operation ).value = pool.getDouble( input
							.readU30() );
					break;

				case Op.PushInt:
					( (PushInt)operation ).value = pool
							.getInt( input.readU30() );
					break;

				case Op.PushNamespace:
					( (PushNamespace)operation ).value = pool
							.getNamespace( input.readU30() );
					break;

				case Op.PushShort:
					( (PushShort)operation ).value = input.readU30();
					break;

				case Op.PushString:
					( (PushString)operation ).value = pool.getString( input
							.readU30() );
					break;

				case Op.PushUInt:
					( (PushUInt)operation ).value = pool.getUInt( input
							.readU30() );
					break;

				case Op.SetLocal:
					( (SetLocal)operation ).register = input.readU30();
					break;

				case Op.SetGlobalSlot:
					( (SetGlobalSlot)operation ).slotIndex = input.readU30();
					break;

				case Op.SetProperty:
					( (SetProperty)operation ).property = pool
							.getMultiname( input.readU30() );
					break;

				case Op.SetSlot:
					( (SetSlot)operation ).slotIndex = input.readU30();
					break;

				case Op.SetSuper:
					( (SetSuper)operation ).property = pool.getMultiname( input
							.readU30() );
					break;
			}

			result.add( operation );
		}

		try
		{
			markers.solve();
		}
		catch( final MarkerException ex )
		{
			new BytecodePrinter( System.out ).interpret( new AbcEnvironment(),
					result );

			throw ex;
		}

		return result;
	}

	public Bytecode decode( final byte[] code ) throws IOException,
			MarkerException
	{
		return decode( new ByteArrayInputStream( code ) );
	}

	public Bytecode decode( final InputStream input ) throws IOException,
			MarkerException
	{
		return decode( new AbcInputStream( input ) );
	}
}

File: apparat-core/src/main/scala/apparat/bytecode/BytecodeDecoder.scala
/*
 * This file is part of Apparat.
 *
 * Copyright (C) 2010 Joa Ebert
 * http://www.joa-ebert.com/
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package apparat.bytecode

import annotation.tailrec
import collection.immutable.{TreeMap, SortedMap}
import collection.mutable.ListBuffer
import java.io.{ByteArrayInputStream => JByteArrayInputStream}
import operations._
import apparat.abc.{AbcMethodBody, AbcInputStream, AbcExceptionHandler, Abc}

object BytecodeDecoder {
	def apply(bytecode: Array[Byte], abcExceptions: Array[AbcExceptionHandler], body: AbcMethodBody)(implicit abc: Abc) = {
		val input = new AbcInputStream(new JByteArrayInputStream(bytecode))
		val cpool = abc.cpool
		val markers = new MarkerManager()
		val exceptions = abcExceptions map {
			handler => new BytecodeExceptionHandler(
				markers.putMarkerAt(handler.from),
				markers.putMarkerAt(handler.to),
				markers.putMarkerAt(handler.target),
				handler.typeName, handler.varName)
		}

		@inline def u08 = input.readU08()
		@inline def s08 = input.readS08()
		@inline def s24 = input.readS24()
		@inline def u30 = input.readU30()
		@inline def s30 = input.readS30()
		@inline def name = cpool.names(u30)
		@inline def string = cpool.strings(u30)
		@inline def numArguments = u30
		@inline def register = u30
		@inline def slot = u30
		@inline def property = name
		@inline def marker(implicit position: Int) = markers putMarkerAt (position + 0x04 + s24)
		@inline def readOp(implicit position: Int): AbstractOp = input.readU08() match {
			case Op.add => Add()
			case Op.add_i => AddInt()
			case Op.add_p => error("add_p")
			case Op.applytype => ApplyType(numArguments)
			case Op.astype => AsType(name)
			case Op.astypelate => AsTypeLate()
			case Op.bitand => BitAnd()
			case Op.bitnot => BitNot()
			case Op.bitor => BitOr()
			case Op.bitxor => BitXor()
			case Op.bkpt => Breakpoint()
			case Op.bkptline => BreakpointLine()
			case Op.call => Call(numArguments)
			case Op.callmethod => CallMethod(u30, numArguments)
			case Op.callproperty => CallProperty(property, numArguments)
			case Op.callproplex => CallPropLex(property, numArguments)
			case Op.callpropvoid => CallPropVoid(property, numArguments)
			case Op.callstatic => CallStatic(abc methods u30, numArguments)
			case Op.callsuper => CallSuper(property, numArguments)
			case Op.callsupervoid => CallSuperVoid(property, numArguments)
			case Op.checkfilter => CheckFilter()
			case Op.coerce => Coerce(name)
			case Op.coerce_a => CoerceAny()
			case Op.coerce_b => CoerceBoolean()
			case Op.coerce_d => CoerceDouble()
			case Op.coerce_i => CoerceInt()
			case Op.coerce_o => CoerceObject()
			case Op.coerce_s => CoerceString()
			case Op.coerce_u => CoerceUInt()
			case Op.construct => Construct(numArguments)
			case Op.constructprop => ConstructProp(property, numArguments)
			case Op.constructsuper => ConstructSuper(numArguments)
			case Op.convert_b => ConvertBoolean()
			case Op.convert_d => ConvertDouble()
			case Op.convert_i => ConvertInt()
			case Op.convert_m => error("convert_m")
			case Op.convert_m_p => error("convet_m_p")
			case Op.convert_o => ConvertObject()
			case Op.convert_s => ConvertString()
			case Op.convert_u => ConvertUInt()
			case Op.debug => Debug(u08, string, u08, u30)
			case Op.debugfile => DebugFile(string)
			case Op.debugline => DebugLine(u30)
			case Op.declocal => DecLocal(register)
			case Op.declocal_i => DecLocalInt(register)
			case Op.declocal_p => error("declocal_p")
			case Op.decrement => Decrement()
			case Op.decrement_i => DecrementInt()
			case Op.decrement_p => error("decrement_p")
			case Op.deldescendants => error("deldescendants")
			case Op.deleteproperty => DeleteProperty(property)
			case Op.divide => Divide()
			case Op.divide_p => error("divide_p")
			case Op.dup => Dup()
			case Op.dxns => DefaultXMLNamespace(string)
			case Op.dxnslate => DefaultXMLNamespaceLate()
			case Op.equals => Equals()
			case Op.esc_xattr => EscapeXMLAttribute()
			case Op.esc_xelem => EscapeXMLElement()
			case Op.finddef => error("finddef")
			case Op.findproperty => FindProperty(property)
			case Op.findpropstrict => FindPropStrict(property)
			case Op.getdescendants => GetDescendants(property)
			case Op.getglobalscope => GetGlobalScope()
			case Op.getglobalslot => GetGlobalSlot(slot)
			case Op.getlex => GetLex(name)
			case Op.getlocal => GetLocal(register)
			case Op.getlocal0 => GetLocal(0)
			case Op.getlocal1 => GetLocal(1)
			case Op.getlocal2 => GetLocal(2)
			case Op.getlocal3 => GetLocal(3)
			case Op.getproperty => GetProperty(property)
			case Op.getscopeobject => GetScopeObject(u08)
			case Op.getslot => GetSlot(slot)
			case Op.getsuper => GetSuper(property)
			case Op.greaterequals => GreaterEquals()
			case Op.greaterthan => GreaterThan()
			case Op.hasnext => HasNext()
			case Op.hasnext2 => HasNext2(register, register)
			case Op.ifeq => IfEqual(marker)
			case Op.iffalse => IfFalse(marker)
			case Op.ifge => IfGreaterEqual(marker)
			case Op.ifgt => IfGreaterThan(marker)
			case Op.ifle => IfLessEqual(marker)
			case Op.iflt => IfLessThan(marker)
			case Op.ifne => IfNotEqual(marker)
			case Op.ifnge => IfNotGreaterEqual(marker)
			case Op.ifngt => IfNotGreaterThan(marker)
			case Op.ifnle => IfNotLessEqual(marker)
			case Op.ifnlt => IfNotLessThan(marker)
			case Op.ifstricteq => IfStrictEqual(marker)
			case Op.ifstrictne => IfStrictNotEqual(marker)
			case Op.iftrue => IfTrue(marker)
			case Op.in => In()
			case Op.inclocal => IncLocal(register)
			case Op.inclocal_i => IncLocalInt(register)
			case Op.inclocal_p => error("inclocal_p")
			case Op.increment => Increment()
			case Op.increment_i => IncrementInt()
			case Op.increment_p => error("increment_p")
			case Op.initproperty => InitProperty(property)
			case Op.instanceof => InstanceOf()
			case Op.istype => IsType(name)
			case Op.istypelate => IsTypeLate()
			case Op.jump => Jump(marker)
			case Op.kill => Kill(register)
			case Op.label => Label()
			case Op.lessequals => LessEquals()
			case Op.lessthan => LessThan()
			case Op.lf32 => GetFloat()
			case Op.lf64 => GetDouble()
			case Op.li16 => GetShort()
			case Op.li32 => GetInt()
			case Op.li8 => GetByte()
			case Op.lookupswitch => LookupSwitch(markers putMarkerAt (position + s24), Array.fill(u30 + 1) { markers putMarkerAt (position + s24) })
			case Op.lshift => ShiftLeft()
			case Op.modulo => Modulo()
			case Op.modulo_p => error("modulo_p")
			case Op.multiply => Multiply()
			case Op.multiply_i => MultiplyInt()
			case Op.multiply_p => error("multiply_p")
			case Op.negate => Negate()
			case Op.negate_i => NegateInt()
			case Op.negate_p => error("negate_p")
			case Op.newactivation => NewActivation()
			case Op.newarray => NewArray(numArguments)
			case Op.newcatch => NewCatch(exceptions(u30))
			case Op.newclass => NewClass(abc types u30)
			case Op.newfunction => NewFunction(abc methods u30)
			case Op.newobject => NewObject(numArguments)
			case Op.nextname => NextName()
			case Op.nextvalue => NextValue()
			case Op.nop => Nop()
			case Op.not => Not()
			case Op.pop => Pop()
			case Op.popscope => PopScope()
			case Op.pushbyte => PushByte(s08)
			case Op.pushdecimal => error("pushdecimal")
			case Op.pushdnan => error("pushdnan")//push decimal nan
			case Op.pushdouble => PushDouble(cpool doubles u30)
			case Op.pushfalse => PushFalse()
			case Op.pushint => PushInt(cpool ints u30)
			case Op.pushnamespace => PushNamespace(cpool namespaces u30)
			case Op.pushnan => PushNaN()
			case Op.pushnull => PushNull()
			case Op.pushscope => PushScope()
			case Op.pushshort => PushShort(s30) // pushshort is signed
			case Op.pushstring => PushString(string)
			case Op.pushtrue => PushTrue()
			case Op.pushuint => PushUInt(cpool uints u30)
			case Op.pushundefined => PushUndefined()
			case Op.pushuninitialized => error("pushuninitialized")
			case Op.pushwith => PushWith()
			case Op.returnvalue => ReturnValue()
			case Op.returnvoid => ReturnVoid()
			case Op.rshift => ShiftRight()
			case Op.setglobalslot => SetGlobalSlot(slot)
			case Op.setlocal => SetLocal(register)
			case Op.setlocal0 => SetLocal(0)
			case Op.setlocal1 => SetLocal(1)
			case Op.setlocal2 => SetLocal(2)
			case Op.setlocal3 => SetLocal(3)
			case Op.setproperty => SetProperty(property)
			case Op.setslot => SetSlot(slot)
			case Op.setsuper => SetSuper(property)
			case Op.sf32 => SetFloat()
			case Op.sf64 => SetDouble()
			case Op.si16 => SetShort()
			case Op.si32 => SetInt()
			case Op.si8 => SetByte()
			case Op.strictequals => StrictEquals()
			case Op.subtract => Subtract()
			case Op.subtract_i => SubtractInt()
			case Op.subtract_p => error("subtract_p")
			case Op.swap => Swap()
			case Op.sxi1 => Sign1()
			case Op.sxi16 => Sign16()
			case Op.sxi8 => Sign8()
			case Op.`throw` => Throw()
			case Op.timestamp => error("timestamp")
			case Op.typeof => TypeOf()
			case Op.urshift => ShiftRightUnsigned()
			case x => error("Unknown opcode " + x + ".")
		}

		@tailrec def build(list: List[AbstractOp], map: SortedMap[Int, AbstractOp]): (List[AbstractOp], SortedMap[Int, AbstractOp]) = input.available match {
			case 0 => (list, map)
			case _ => {
				val pos = input.position
				val op = readOp(pos)

				build(op :: list, map + (pos -> op))
			}
		}

		try {
			val (ops, map) = build(Nil, TreeMap())
			new Bytecode(ops.reverse, markers solve map, exceptions, Some(body))
		}
		finally {
			try { input.close() } catch { case _ => {} }
		}
	}
}
